week-5-serving-streamlit/app.py

```diff
@@ -0,0 +1,70 @@
+import os
+from typing import List, Text, Tuple
+from pathlib import Path
+import torch
+from transformers import BertTokenizer
+from model import RegressionModel
+import gdown
+import streamlit as st
+
+
+MODEL_URL = "https://drive.google.com/file/d/1M5SFB5cYS7Q3oLpkEQVGERXtquwPxury/view?usp=sharing"
+MODEL_SAVE_PATH = "./weights/bert-base-uncased-regression-weights.pth"
+
+
+@st.cache
+def init_model() -> Tuple[torch.nn.Module, torch.nn.Module, torch.device]:
+    folder = Path("./weights")
+    weights = Path(MODEL_SAVE_PATH)
+    if not folder.exists():
+        os.mkdir("./weights")
+    if not weights.exists():
+        gdown.download(
+            fuzzy=True, 
+            url=MODEL_URL,
+            output=MODEL_SAVE_PATH)
+    device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
+    print(f">>> using device: {device}")
+    model = RegressionModel()
+    model.load_state_dict(torch.load(MODEL_SAVE_PATH))
+    model.to(device)
+    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
+    model.eval()
+    return model, tokenizer, device
+
+
+@st.cache
```

Comment:
why do we need cache where?

Reply:
For inference it is good idea to cache predictions especially in the case of heavy model.